### PR TITLE
LinearDensity now working with residues/segments

### LINKED ISSUE #3571: LinearDensity does not work on residues or segments
## Expected behavior ##

LinearDensity can calculate mass and charge densities for groups of atoms, residues, segments and fragments

## Actual behavior ##

Running LinearDensity with grouping='residues' or grouping='segments' causes AttributeError
AttributeError: Residue has no attribute centroid.
AttributeError: Segment has no attribute centroid.

## Code to reproduce the behavior ##

``` python
import MDAnalysis as mda
from MDAnalysis.tests.datafiles import waterPSF, waterDCD
from MDAnalysis.analysis.lineardensity import LinearDensity

u = mda.Universe(waterPSF, waterDCD)
selection = u.select_atoms("all")

ld = LinearDensity(selection, grouping="residues").run()    # Fails
ld = LinearDensity(selection, grouping="segments").run()    # Fails

```

## Current version of MDAnalysis ##

- Which version are you using?  2.2.0-dev0
- Which version of Python (`python -V`)? 3.9.10
- Which operating system? Ubuntu 18.04

### PR DESCRIPTION
**Pull request by BFedder**
Fixes #3571

LinearDensity did not work with grouping='residues' and grouping='segments'.
This is now fixed, and more tests have been added.

Changes made:
MDAnalysis/analysis/lineardensity.py:
  - in _prepare(self):
    - try/except statement replaced by if statements checking the grouping
      specified for LinearDensity.
    - masses and charges are obtained in a way appropriate for each case.

  - in _single_frame(self):
    - 'positions' variable assignment changed from elem.centroid() to
      elem.atoms.centroid() so it works for residues/segments

  - in _conclude(self):
    - Floating point arithmetic issues could lead to negative radicands. A check
      was implemented to set numbers which are close to zero to precisely zero
      before the square root (for standard deviation) is calculated.

MDAnalysisTests/analysis/test_lineardensity.py:
  - renamed previous test function and added expected masses and charges
  - added test functions for the other groupings with their respective expected
    masses and charges

Regarding my suggested change to _conclude(self): 
I stuck with binsize=5 for the new tests for the other groupings for internal 
consistency in the tests. With this bin size, the positions of the centres of 
geometry are invariant, yielding standard deviations equal to zero. Floating point
imprecision made it possible for negative values on the order of e-13 to e-14 to 
appear and cause a warning from np.sqrt() though, and that is why I added the 
extra step of calculating the radicand and editing it where appropriate beforehand. 
Other solutions are to just ignore this and accept nan values in the result arrays, 
or change the bin size for the tests. I was hesitant to do the later to stay consistent
with the established test, but I'm happy to change this of course if that's the consensus.

PR Checklist
------------
 - [ ] Tests?
 - [ ] Docs?
 - [ ] CHANGELOG updated?
 - [ ] Issue raised/referenced?

### CODE REVIEW

File: package/MDAnalysis/analysis/lineardensity.py
Comment (IAlibay) [inline package/MDAnalysis/analysis/lineardensity.py:161]: rather than using 3 `if` it might be worth making this `if/elif/else`? Possibly throwing an error if you have an unknown grouping?

File: package/MDAnalysis/analysis/lineardensity.py
Comment (IAlibay) [inline package/MDAnalysis/analysis/lineardensity.py]: So someone like @richardjgowers is probably going to tell me off for this, but the cost of accessing `.atoms.total_mass()` on an atomgroup is pretty much within error as calling `.total_mass()`.

Would it not make sense here to just use the same logic for residues, segments and fragments?

File: package/MDAnalysis/analysis/lineardensity.py
Comment (IAlibay) [inline package/MDAnalysis/analysis/lineardensity.py]: unnecessary blank line addition?

File: package/MDAnalysis/analysis/lineardensity.py
Comment (IAlibay) [inline package/MDAnalysis/analysis/lineardensity.py]: pep8 - single blank line between class methods?

**Review by IAlibay — CHANGES_REQUESTED** [review]
Couple of quick comments on the first half of the code (didn't get pass _conclude for now)

**Comment by BFedder** [comment]
Hi, and thanks for the feedback!
First off, I'll be better about checking PEP8 compliance before opening a PR next time... Sorry! 

I could change the code you mention above to the following to have if/elif/else, raise an exception in the case of unknown groupings, and use the same logic for residues, segments and fragments?

``` python
# Get masses and charges for the selection
if self.grouping == "atoms":
    self.masses = self._ags[0].masses
    self.charges = self._ags[0].charges

elif self.grouping in ["residues", "segments", "fragments"]:
    self.masses = np.array([elem.atoms.total_mass() for elem in group])
    self.charges = np.array(
        [elem.atoms.total_charge() for elem in group])

else:
    raise RuntimeError(
        f"{self.grouping} is not a valid value for grouping.")

```

**Comment by IAlibay** [comment]
> First off, I'll be better about checking PEP8 compliance before opening a PR next time... Sorry!

Flake8 can be really useful here, especially if combined with something like VS code. That being said, even of the coredevs are terrible at keeping up with pep8 🙄 

> I could change the code you mention above to the following to have if/elif/else, raise an exception in the case of unknown groupings, and use the same logic for residues, segments and fragments?

Sure, at a quick glance that seems reasonable, although you might want to use an AttributeError instead of a RuntimeError here.

**Comment by IAlibay** [comment]
I'm going to cycle the PR (opening and closing it), I'm not sure why gh actions is being silly today...

**Comment by IAlibay** [comment]
well at least the webhooks got triggered, that should be enough for now until actions comes back (another day another gh service issue... https://www.githubstatus.com/)

File: package/MDAnalysis/analysis/lineardensity.py
Comment (RMeli) [inline package/MDAnalysis/analysis/lineardensity.py]:
```suggestion
            radicand_pos = np.where(np.isclose(radicand_pos, 0), 0, radicand_pos)
```
or something on those lines.

File: package/MDAnalysis/analysis/lineardensity.py
Comment (RMeli) [inline package/MDAnalysis/analysis/lineardensity.py]:
```suggestion
            radicand_char = np.where(np.isclose(radicand_char, 0), 0, radicand_char)
```
or something on those lines.

File: package/MDAnalysis/analysis/lineardensity.py
Comment (RMeli) [reply package/MDAnalysis/analysis/lineardensity.py]: Or even
```suggestion
            radicand_pos[radicand_pos < 0] = 0
```

File: package/MDAnalysis/analysis/lineardensity.py
Comment (RMeli) [reply package/MDAnalysis/analysis/lineardensity.py]: Or even
```suggestion
            radicand_char[radicand_char < 0] = 0
```

File: package/MDAnalysis/analysis/lineardensity.py
Comment (BFedder) [reply package/MDAnalysis/analysis/lineardensity.py]: Thanks @RMeli! I was sure there must be a better solution for this and your suggestion looks much cleaner. 
I think using your first suggestion with `np.where` has two advantages: 

- If an actual negative number that's not close to zero shows up at this point, that's an error that should be flagged up/ throw a warning
- positive numbers on the order of e-14 are not meaningful results and should be 0 as well

But your second suggestion looks like it would give better performance. I'm not sure how much of a difference this would make, could people weigh in here? If the first suggestion is not much slower than the second, than that's the one I would go for.

Comment (RMeli) [reply package/MDAnalysis/analysis/lineardensity.py]: Yes, precisely. The first option is safer, the second one is faster. 
```
In [1]: import numpy as np

In [2]: x = np.array([1.0, 1e-10, -1e-10, 0.1])

In [3]: %timeit y = np.where(np.isclose(x, 0), 0, x)
40.4 µs ± 325 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

In [4]: z = x.copy()

In [5]: %timeit z[z < 0] = 0
2.1 µs ± 17.1 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
```

```
In [6]: x = np.random.normal(size=10000)

In [7]: %timeit y = np.where(np.isclose(x, 0), 0, x)
90.5 µs ± 1.23 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

In [8]: z = x.copy()

In [9]: %timeit z[z < 0] = 0
6.72 µs ± 65.2 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
```

In scientific software, IMHO, correctness is far more important than speed. However, since this is the `_conclude` function, if the code executed before performs enough checks it might be safe to use the second option here (i.e. the only negative values only come from numerical errors).

Comment (IAlibay) [reply package/MDAnalysis/analysis/lineardensity.py]: Because we only have to do this once, I might vote for the safer option here, unless we can guarantee that negative numbers are always floating point issues (I've not really delved into what the cause is here, maybe that is the case).

Comment (BFedder) [reply package/MDAnalysis/analysis/lineardensity.py]: Thanks for this clear breakdown, RMeli! 

IAlibay, I have looked into the code again and I can't see a way for actual negative numbers to appear here. I would suggest the following, then:

`radicand_pos[radicand_pos < 1e-9] = 0`

This is fast and also takes care of floating point errors that give a tiny positive number. 

`x = np.array([1.0, 1e-10, -1e-10, 0.1])`
`%timeit x[x < 1e-9] = 0`
`1.07 µs ± 8.22 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)`

Comment (IAlibay) [reply package/MDAnalysis/analysis/lineardensity.py]: Yeah sure I can live with that

Comment (BFedder) [reply package/MDAnalysis/analysis/lineardensity.py]: Changed now.

Comment (orbeckst) [reply package/MDAnalysis/analysis/lineardensity.py]: Sorry for the drive-by comment: If you're only concerned with negative numbers then I'd keep @RMeli 's suggestion with `x[x<0]` instead of introducing an arbitrary cutoff (and 1e-9 is not necessarily small). Numerical roundoff is normal and if anything I'd want to take care of it at the very end of a calculation and not infer in the middle if something ought to be zero or not. The objective of "no negative numbers" is clear. The objective of "should be zero" is not. Thus I only would implement the first one.

File: package/MDAnalysis/analysis/lineardensity.py
Comment (PicoCentauri) [inline package/MDAnalysis/analysis/lineardensity.py]: The method `group.total_mass` and `total_charge` have a parameter for `compound` to provide grouping which is faster than list compregensions. Is there a reason not to use it here?

File: package/MDAnalysis/analysis/lineardensity.py
Comment (PicoCentauri) [inline package/MDAnalysis/analysis/lineardensity.py]: The center of geometry has no physical meaning. I would still stick to the center of mass. You can get it with 
`positions  = group.center_of_mass(compound=self.grouping)`

File: testsuite/MDAnalysisTests/analysis/test_lineardensity.py
Comment (PicoCentauri) [inline testsuite/MDAnalysisTests/analysis/test_lineardensity.py]: You may simplify duplicated code with `pytest.mark.parametrize`. Look for examples in the testsuite, there are plenty of them. 

Additionally, there are no tests for the actual calculated density. I would at least add a check for the mean of `results.x.pos` and `results.x.char`

**Review by PicoCentauri — CHANGES_REQUESTED** [review]
Good work to fix this!

I have some code simplifications and one physical question/addition. For me the center of geometry does not make sense to use as a reference point. It has no physical meaning.

File: package/MDAnalysis/analysis/lineardensity.py
Comment (PicoCentauri) [reply package/MDAnalysis/analysis/lineardensity.py]: I am confused by his operation. Do have an example where this happens? I never had an issue with `nan` in calculating density profiles.

Edit: Actually, I also had them. I [usually](https://gitlab.com/maicos-devel/maicos/-/blob/main/maicos/modules/density.py#L358-360) convert negative number with `np.nan_to_num` to 0 after calculating the standard deviation.

File: package/MDAnalysis/analysis/lineardensity.py
Comment (IAlibay) [inline package/MDAnalysis/analysis/lineardensity.py:170]: Make sure this is covered by tests please.

File: package/MDAnalysis/analysis/lineardensity.py
Comment (IAlibay) [reply package/MDAnalysis/analysis/lineardensity.py]: Seems like one of those where the original code thought COM but actually attempted (and failed) to do COG.

I agree with COM here too.

This does need to be clearly signposted in the documentation for this class. Can you see if you could add this somewhere in the docstring, and also add a relevant example of using this for there various groupings too please?

**Review by IAlibay — CHANGES_REQUESTED** [review]

File: package/MDAnalysis/analysis/lineardensity.py
Comment (BFedder) [reply package/MDAnalysis/analysis/lineardensity.py]: That's a good point, @orbeckst, thanks! I have changed this to `x[x<0]` now. 

@PicoCentauri, Ah, I did not know about `np.nan_to_num`, thank you! For now, I stuck with the other suggestion because it is slightly faster, but is there another advantage to nan_to_num?

File: package/MDAnalysis/analysis/lineardensity.py
Comment (BFedder) [reply package/MDAnalysis/analysis/lineardensity.py]: No reason that I can see, thanks for this suggestion!

File: package/MDAnalysis/analysis/lineardensity.py
Comment (BFedder) [reply package/MDAnalysis/analysis/lineardensity.py]: Yes, I went with COG because that's what was used previously. I have now changed it to COM and made use of `compound`, thanks. 

@IAlibay, I have changed the class docstring in the Parameters section and in the version history, but I am not sure what a useful example would be, because the user input has not changed and the output for the relevant groupings is only now available, so it should not break any pre-existing scripts. Could you please clarify what examples could help here?

**Review by PicoCentauri — APPROVED** [review]
Very well. LGTM

File: package/MDAnalysis/analysis/lineardensity.py
Comment (PicoCentauri) [reply package/MDAnalysis/analysis/lineardensity.py]: If it is faster it is always better. I am not aware of any other advantages compared to your solution.

File: package/MDAnalysis/analysis/lineardensity.py
Comment (IAlibay) [reply package/MDAnalysis/analysis/lineardensity.py]: Just the same as what is already under example but saying something along the lines of "if you wanted to do this per residue you could do it this way <code block>". It's a tiny addition but it can be quite a nice thing to read if you're not super sure how the grouping argument works.

File: testsuite/MDAnalysisTests/analysis/test_lineardensity.py
Comment (BFedder) [reply testsuite/MDAnalysisTests/analysis/test_lineardensity.py]: I have now refactored the test suite to limit duplicate code, thanks for the suggestion. 

The reason I had not previously added tests for the calculated densities is that I am not sure how I would calculate these on my own. For now, I have taken the output of lineardensity.py and used that as placeholder tests, but that obviously does not do much in the way of actual testing. Do you have some advice on how I could get the "ground truth" data for the densities?

Comment (PicoCentauri) [reply testsuite/MDAnalysisTests/analysis/test_lineardensity.py]: You are right. The current values are just magic numbers. These will probably fail when we do something with the code. Depending on how much effort you want to invest I propose the following:

Generate a trajectory based on random positions of particles in a cubic box. Draw the positions from a uniform distribution. Run the analysis and compare the densities. You could do three kinds of system: uncharged particles, charged particles and maybe charged dimers. This is probably beyond the scope of this PR but would be a valuable addition!

File: package/MDAnalysis/analysis/lineardensity.py
Comment (BFedder) [reply package/MDAnalysis/analysis/lineardensity.py]: I have added an example of this now

File: testsuite/MDAnalysisTests/analysis/test_lineardensity.py
Comment (BFedder) [reply testsuite/MDAnalysisTests/analysis/test_lineardensity.py]: One reason why I stuck with binsize=5 with my new tests is that the original test, which also used this value, had expected values for `results.x.pos`. I thought that this would at least be partial validation for the restored functionality. But independently determined expected values are definitely still needed here. 

Thanks for the suggestion! For now, I need to focus on my GSoC application/proposal, but I would be interested in working on LinearDensity further. I originally started looking into this module because it looked like a good candidate for more tests on codecov, and that's when I encountered the bug fixed in this PR. It looks like between adding tests for the parallel analysis code, getting expected values for the tests, and issues #2507 and #2508, there is still quite a bit to be done here, and I'd be happy to look into it more after more work on my application. For now, I can open an issue describing the need for more test data, and this PR could get merged?

Comment (PicoCentauri) [reply testsuite/MDAnalysisTests/analysis/test_lineardensity.py]: Yes take your time but opening an issue is a good idea!

Comment (BFedder) [reply testsuite/MDAnalysisTests/analysis/test_lineardensity.py]: I have now opened issue #3603 for improved tests.

**Comment by orbeckst** [comment]
@BFedder can you please resolve the conflicts?

**Comment by orbeckst** [comment]
@IAlibay can you please check if your comments were addressed?

File: testsuite/MDAnalysisTests/analysis/test_lineardensity.py
Comment (orbeckst) [reply testsuite/MDAnalysisTests/analysis/test_lineardensity.py]: Just a quick comment: "ground truth" tests are excellent when you can get them. However, "regression tests" are the next best thing: they will check that future changes do not change what was once considered working. For this PR I would be totally happy with regression tests, especially as the original code also used them.

Thanks for raising the issue, that's a good approach to doing incremental software development: do small working units and keep track of what needs to be done through issues.

File: package/MDAnalysis/analysis/lineardensity.py
Comment (IAlibay) [inline package/MDAnalysis/analysis/lineardensity.py]:
```suggestion

    Alternatively, the other types of groupings can be selected using the ``grouping``
    keyword. For example to calculated the density based on the :class:`ResidueGroup`s
    of the system:
    
    .. code-block:: python
       ldens = LinearDensity(selection, grouping='residues', binsize=1.0)
       ldens.run()

```

File: testsuite/MDAnalysisTests/analysis/test_lineardensity.py
Comment (IAlibay) [inline testsuite/MDAnalysisTests/analysis/test_lineardensity.py]: PEP8 is a suggestion not an obligation - don't use this against me

```suggestion
@pytest.mark.parametrize("grouping, expected_masses, expected_charges, expected_xpos, expected_xchar", [
    ("atoms", expected_masses_atoms, expected_charges_atoms,
     expected_xpos_atoms, expected_xchar_atoms),
    ("residues", expected_masses_residues, expected_charges_residues,
     expected_xpos_residues, expected_xchar_residues),
    ("segments", expected_masses_segments, expected_charges_segments,
     expected_xpos_segments, expected_xchar_segments),
    ("fragments", expected_masses_fragments, expected_charges_fragments,
     expected_xpos_fragments, expected_xchar_fragments)
])
```

**Review by IAlibay — CHANGES_REQUESTED** [review]
Just two small suggested changes - I'm directly suggesting them here since they are more "my style" than anything. Feel free to say no, I'll approve after.

**Comment by BFedder** [comment]
Thanks, I have taken the suggestions on board

**Review by IAlibay — APPROVED** [review]
Thanks @BFedder 

@MDAnalysis/gsoc-mentors - I can't merge this as a GSoC PR due to conflicts of interest. Can someone please take over the responsibility of merging this?

**Comment by orbeckst** [comment]
@PicoCentauri can you please have a final look and then squash-merge? Thanks.

**Review by PicoCentauri — APPROVED** [review]

**Comment by PicoCentauri** [comment]
Thanks @BFedder for your fix!

**Pull request merged**
